snapshot/snapshotfs/upload.go

```diff
@@ -255,29 +255,52 @@ func (u *Uploader) uploadDir(ctx context.Context, rootDir fs.Directory, policyTr
 	return de, err
 }
 
-func (u *Uploader) foreachEntryUnlessCancelled(relativePath string, entries fs.Entries, cb func(entry fs.Entry, entryRelativePath string) error) error {
-	for _, entry := range entries {
-		if u.IsCancelled() {
-			return errCancelled
-		}
+func (u *Uploader) foreachEntryUnlessCancelled(ctx context.Context, parallel int, relativePath string, entries fs.Entries, cb func(ctx context.Context, index int, entry fs.Entry, entryRelativePath string) error) error {
+	type entryWithIndex struct {
+		entry fs.Entry
+		index int
+	}
 
-		entryRelativePath := relativePath + "/" + entry.Name()
+	ch := make(chan entryWithIndex)
+	eg, ctx := errgroup.WithContext(ctx)
 
-		if err := cb(entry, entryRelativePath); err != nil {
-			return err
+	// one goroutine to pump entries into channel until ctx is closed.
+	eg.Go(func() error {
+		defer close(ch)
+
+		for i, e := range entries {
+			select {
+			case ch <- entryWithIndex{e, i}: // sent to channel
+			case <-ctx.Done(): // context closed
+				return nil
+			}
 		}
-	}
+		return nil
+	})
 
-	return nil
-}
+	// launch N workers in parallel
+	for i := 0; i < parallel; i++ {
+		eg.Go(func() error {
+			for ewi := range ch {
+				if u.IsCancelled() {
+					return errCancelled
+				}
+
+				entryRelativePath := relativePath + "/" + ewi.entry.Name()
+				if err := cb(ctx, ewi.index, ewi.entry, entryRelativePath); err != nil {
+					return err
+				}
+			}
 
-type entryResult struct {
-	err error
-	de  *snapshot.DirEntry
+			return nil
+		})
+	}
+
+	return eg.Wait()
 }
 
 func (u *Uploader) processSubdirectories(ctx context.Context, relativePath string, entries fs.Entries, policyTree *policy.Tree, previousEntries []fs.Entries, dirManifest *snapshot.DirManifest, summ *fs.DirectorySummary) error {
-	return u.foreachEntryUnlessCancelled(relativePath, entries, func(entry fs.Entry, entryRelativePath string) error {
+	return u.foreachEntryUnlessCancelled(ctx, 1, relativePath, entries, func(ctx context.Context, index int, entry fs.Entry, entryRelativePath string) error {
```

Comment:
This is brittle. The update in line [349](https://github.com/kopia/kopia/pull/331/commits/a7ed40fb05e7b729259a0a8b8f62da584d946143#diff-d711584d735b389e3d0f01af1055bc6bR349) works because `parallel=1`, but would blow otherwise.

Reply:
added constant to make this more obvious, we can work on making this actually parallelizable later